### ISSUE TITLE
[OJS] pdfJsViewer rendering problems with Safari 10.1.

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2557]
Hi,

A journal editor contacted me a month ago concerning font problems in pdf files. Originally it seemed that it was a problem with the new Adobe CC suite, but after talking with Adobe and Apple the problem ended up being with the pdf.js library used in OJS, or actually a bug in Safari.

This is a known issue for pdf.js developers, but worth watching if it is resolved soon so that the OJS plugin can be upgraded to the fixed version of pdf.js

https://github.com/mozilla/pdf.js/issues/8231

**Comment by Ph-We** [comment]
Hi @ajnyga, We had several reported problems of this sort with Windows too.

**Comment by ajnyga** [comment]
I have hear that as well, especially with Win10 + Firefox. However, I can not reproduce this with Wint7 + Firefox, Chrome and IE or with OSX + Firefox and Chrome. Only with OSX + Safari. The thread at pdf.js repository suggests that this is only a Safari issue, but yes, there is more to it.

**Comment by Ph-We** [comment]
Our users reported both the issue with incorrect fonts rendering (OSX, Windows) and incorrect fonts display for certain zoom levels (blurred). The latter I could reproduce in Windows only:  https://github.com/mozilla/pdf.js/issues/7630

Some of our journals had to turn off pdf.js completely.

**Comment by asmecher** [comment]
> Some of our journals had to turn off pdf.js completely.

Hmm, I'm sorry to hear this -- we used to use https://pdfobject.com/ to embed plugins but for a while that was unmaintained, and besides, it presented very inconsistently on different systems and didn't permit e.g. JS tools to interact with the presented content (think Hypothes.is embedding on a PDF view). Moving to PDF.js was a huge improvement and I was hoping it would satisfy everyone. Are you tracking these issues in the PDF.js issue list for progress?

**Comment by Ph-We** [comment]
I was really surprised that it could not render correctly even those PDFs, which were prepared with TEX((
Some possible alternatives: http://www.jqueryrain.com/2012/09/best-jquery-pdf-viewer-plugin-examples/

**Comment by ajnyga** [comment]
I am a bit sceptical whether changing the library would help, but it would be fairly easy to do an alternative pdf plugin based on one of the listed viewers there. Probably not many lines would have to change.

I am tracking the issue above. From what I have seen this could a be combination of used fonts, pdf version and browser. We are not getting that many reports about this, but as I said, the problem is there with some pdf files and I can reproduce it with Safari.

I would be interested to hear what people are seeing with  Windows 10 + Firefox. I mean, I guess there should be no difference, but I have heard more than once that windows users with ff are also having problems.

**Comment by asmecher** [comment]
Agreed, additional plugins would be easy to write. However, we used to support PDFObject (built in), PDF.js (plugin), and Google Viewer (plugin), but droped everything but PDF.js because it was so superior to the others.

**Comment by ajnyga** [comment]
Hi @asmecher, seems like there could be a fix for this now: https://github.com/box/box-content-preview/pull/332

**Comment by asmecher** [comment]
Thanks, @ajnyga -- I see those changes are approved but not merged, so I'll schedule this issue for 3.1 and hopefully it'll get included before that has to get wrapped up. Then we can update the version of PDF.js included in OJS before the release is built.

**Comment by ajnyga** [comment]
Hi, I think that it does not fix PDF.js itself, but a way to get around the problem.

They are basically checking if the user is using Safari on then changing the value of `PDFJS.disableFontFace`.  Not sure what that does exactly.

**Comment by asmecher** [comment]
Ah, I see -- PDF.js seems to be tracking this at https://github.com/mozilla/pdf.js/issues/8231. There's a link there to a Ghostscript command line that'll downgrade the PDF to a lower version and thus work around the issue that way. I'd like to avoid patching/tweaking upstream code in favour of watching for the upstream fix.

**Comment by ajnyga** [comment]
Yes I agree, PDF.js has to solve this someway in the near future. Probably there will be a lot more people seeing this issue when more pdf's with newer versions are created and used with pdf.js.

**Comment by ajnyga** [comment]
New comment:"Note that this issue appears to be resolved in Safari 11."

But, I am pretty sure that one of our editors was seeing this in Firefox as well...

**Comment by asmecher** [comment]
Not sure it'll fix this, but I did just commit updates to the PDF.js library to bring it to release 1.8.188:

- https://github.com/pkp/omp/commit/94869614df6e37cc001cf6615b7eaf09f73de36c (for OMP)
- https://github.com/pkp/pdfJsViewer/commit/72719cc67eb5af591493179f324a8c82ab31bf25 (for OJS)

**Comment by NateWr** [comment]
@ajnyga Would you have an opportunity to test this week? If not, that's ok, but I may defer this issue to test at a later date since 3.1 is almost out the door.

**Comment by ajnyga** [comment]
Hi,

I am not sure if it is Safari 11 or what, but I can not reproduce the problem at the moment. I asked the editor to check two formerly problematic pdf-files on my test server and will let you know what she says. This is definitely an issue that is sensitive for certain browsers/versions.

**Comment by asmecher** [comment]
Moving off the 3.1 release for later consideration/work/testing.

**Comment by ajnyga** [comment]
Sorry I forgot about this. 

So I installed the latest pdf.js plugin (1.8.188) to my test server and added a couple of the problematic pdf's. The editor went to check those and this is the outcome:

OS X + Safari 10.1.2 => Issue still occurs
Ipad iOS + Safari 10.3.3 => Issue still occurs
Ipad iOS + Safari 10.3.2 => Issue does not occur !!!

My earlier experience and also mentioned in the other threat is that with Safari 11 this also works.
I also have an old iPad with Safari 8, and that also works.

So basically the new version of pdf.js does not make any difference here. 
The problem seems to be with Safari and to be exact with Safari version 10.3.3. (iOS) and Safari 10.1 in OS X.

**Comment by asmecher** [comment]
Hmm, I'm still not sure there's anything we can do here.

**Comment by ajnyga** [comment]
No we cant unless some other pdf library magically appears. So depending on how you like to keep on tracking this issue go ahead and close it or just leave it open.

**Comment by asmecher** [comment]
@ajnyga, I suspect it'll either get fixed or it won't :)

To make that more helpful, what I mean is that we'll either see the problem go away with an upstream pdf.js update, or that particular version of Safari will become uncommon and we won't hear anything more about it.

Currently our version of pdf.js is manually maintained in the https://github.com/pkp/pdfJsViewer repo; dependencies in plugins are currently maintained using Composer, but I don't see a good candidate Composer package for pdf.js. It would be easier for us to keep pdf.js up to date if it wasn't being manually maintained. The git repository is way too source-heavy and we just need the built package. @NateWr, are any of our new JS-land tools good candidates for maintaining the pdfJsViewer plugin's copy of PDF.js from upstream sources?

**Comment by NateWr** [comment]
We can use npm to pull it in. Looks like there's a distribution copy here:

https://www.npmjs.com/package/pdfjs-dist

But that would still require us to "build" it and load it, which might not be the worse thing in the world. I'll take a look.

**Comment by NateWr** [comment]
Closing as wontfix since there doesn't seem to be much we can do here.
